### DISCUSSION TITLE
GroupBy support in Linq queries

### DISCUSSION THREAD

**Discussion by jeremydmiller** [JasperFx/marten#2119] [Ideas]
No real plans about how far to go with this yet or timing.

**Comment by AKlaus** [comment]
`Queryable.GroupBy()` throws an error:
```
Could not determine an 'id/Id' field or property for requested document type System.Linq.IGrouping`2
   at Marten.Schema.DocumentMapping.Validate()
   at Marten.Storage.DocumentTable..ctor(DocumentMapping mapping)
   at Marten.Schema.DocumentMapping.<toSchemaObjects>d__129.MoveNext()
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Marten.Storage.Tenant.generateOrUpdateFeature(Type featureType, IFeatureSchema feature)
   at Marten.Storage.Tenant.ensureStorageExists(IList`1 types, Type featureType)
   at Marten.Linq.MartenQueryable`1.executeAsync[TResult](Func`2 source, CancellationToken token)
   at Marten.Linq.MartenQueryable`1.ToListAsync[TResult](CancellationToken token)
   at Marten.QueryableExtensions.ToListAsync[T](IQueryable`1 queryable, CancellationToken token)
```

Therefore:
- Could you please confirm that there is no option of using `GROUP BY` except manually writing an SQL query?
- Are there plans to implement the functionality? (looks unlikely, as it's marked for 1.3 milestone and the current version is 2.4).
- RavenDb provides this functionality via [Map-Reduce indexes](https://ravendb.net/docs/article-page/4.0/http/indexes/map-reduce-indexes). Is there a different (or recommended) approach to reduce data by grouping on a field set?

~~P.S. I understand that Marten is an open source project and public contribution is more than welcome, but I was very surprised to see lack of grouping functionality.~~

**Comment by jeremydmiller** [comment]
@AKlaus You're commenting on the issue in the backlog to add this feature, it's not in the docs, and a very quick search would have shown you that it doesn't appear in any unit test in the codebase.

So,

1. Marten does not yet support GroupBy in Linq
2. Yes, obviously by the existence of this issue
3. Good for them, and no, not yet

"I was very surprised to see lack of grouping functionality" -- please don't ever use a phrase like that in an OSS issue. 

It's not already there because it's very hard and time consuming to do from the Linq perspective and not enough folks have wanted it yet. Unlike RavenDb though, you can always revert to Sql and that's the only answer I've got for you here.

**Comment by AKlaus** [comment]
Thank you for your prompt answer :+1:. I agree with everything you said.
It made me think whether I'm solving the right problem if usage of GroupBy is a rare requirement on a NoSQL DB.

**Comment by eouw0o83hf** [comment]
In response to "not enough folks have wanted it yet," I'd like to toss my hat in the ring for needing `GroupBy()` linq functionality. Definitely understand low-priority-ness.

**Comment by robertmaidlaAW** [comment]
Checking in on whether this has been added to the to-do list :P

**Comment by oskardudycz** [comment]
@robertmaidlaAW it's still on the TODO list, although to be fair - not in the top of the priorities (because of our capacity). Next biggest release will be probably mostly focused on Event Store enhancements.

**Comment by igor-maslennikov** [comment]
hello, just checking - any plans to have it covered in the nearest future?

**Reply by tomitesh** [reply]
When is it planned? eagerly waiting for this feature (group by operator)

**Comment by SnakeA92** [comment]
any updates on this ?

**Reply by jeremydmiller** [reply]
Sorry, I haven't thought about this in ages. Hasn't come up all that often. Like I've said time and time again, it's a very expensive thing to build. If you want it though, this: https://github.com/sponsors/JasperFx or anyone who wants to engage JasperFx Software to build it out, feel free to reach out.

Obviously take a PR for that, but it's not a quick "oh, this will be fun" kind of thing to do, and it'll 100% be a permutation hell problem

**Comment by jeremydmiller** [comment]
To be much nicer about this folks, you can do this *outside* of LINQ by dropping down to SQL with the `IQuerySession.Query<T>(sql)` or `IQuerySession.AdvancedSql` APIs. Honestly, I don't think this is likely to be inside of the LINQ support anytime soon, and pure SQL does work. Easier now that you can use JsonPath expressions in PostgreSQL
